File: 2023/day20.nim
import std/algorithm
import std/bitops
import std/deques
import std/enumerate
import std/heapqueue
import std/math
import std/options
import std/sequtils
import std/sets
import std/strscans
import std/strutils
import std/sugar
import std/tables
import std/terminal
import std/times

proc dbg[T](v: T): T = (echo(v); v)

func empty[T](v: seq[T]): bool = v.len == 0

proc split[T](v: seq[T], pred: proc(elt: T): bool {.noSideEffect.}): seq[seq[T]] =
  var current: seq[T]
  for elt in v:
    if pred(elt):
      result.add current
      current = @[]
    else:
      current.add elt
  result.add current

func join[T](parts: seq[seq[T]], sep: T): seq[T] =
  if parts.len > 0:
    result &= parts[0]

  for part in parts[1..^1]:
    result.add sep
    result &= part

###################

type Pulse = enum
  High
  Low

# Something about how I'm defining these objects and methods is very wrong.
#
# Depending on how much of `pushButton` I implement, I either get a type
# mismatch when trying to resolve `mm.send()` below or an error message from
# `gcc` about `Sup` in something that is not a structure or a union.
#
# Neither of these seem like fun to debug right now.

type
  Module = ref object of RootObj
    name: string
    dests: seq[string]
  Broadcaster = ref object of Module
  FlipFlop = ref object of Module
    state: Pulse
  Conjunction = ref object of Module
    state: Table[string, Pulse]

func parseModule(line: string): Module =
  let parts = line.strip.split(" -> ", 1)
  let (name, dests) = (parts[0], parts[1])

  result.dests = dests.split(", ")

  if name[0] == '%':
    result = FlipFlop(name: name[1..^1], state: Low)
  elif name[0] == '&':
    result = Conjunction(name: name[1..^1])
  elif name == "broadcaster":
    result = Broadcaster(name: name)
  else:
    raise newException(ValueError, line)

type Machine = tuple
  modules: Table[string, Module]

func newMachine(text: string): Machine =
  for m in text.strip.splitLines.map(parseModule):
    result.modules[m.name] = m

  for (srcName, srcMod) in result.modules.pairs:
    for dstName in srcMod.dests:
      let dstMod = result.modules[dstName]
      if typeof(dstMod) is Conjunction:
        Conjunction(dstMod).state[srcName] = Low

type Wire = tuple
  src: string
  dest: string
  value: Pulse

method send(m: ref var Module, source: string, p: Pulse): seq[Wire] {.base.} =
  raise newException(CatchableError, "abstract method")

method send(b: ref var Broadcaster, _: string, p: Pulse): seq[Wire] =
  let src = b.name
  for dst in b.dests:
    result.add (src, dst, p)

method send(c: ref var Conjunction, source: string, p: Pulse): seq[Wire] =
  c.state[source] = p

  var output = Low
  for v in c.state.values:
    if v == Low:
      output = High
      break

  let src = c.name
  for dst in c.dests:
    result.add (src, dst, output)

method send(f: ref var FlipFlop, source: string, p: Pulse): seq[Wire] =
  if p == High:
    return @[]

  var output = Low
  case f.state:
    of High:
      f.state = Low
      output = Low
    of Low:
      f.state = High
      output = High

  let src = f.name
  for dst in f.dests:
    result.add (src, dst, output)

proc pushButton(m: var Machine): CountTable[Pulse] =
  var queue: Deque[Wire]

  queue.addLast ("button", "broadcaster", Low)

  while queue.len > 0:
    let (src, name, pulse) = queue.popFirst
    result.inc pulse

    let mm = m.modules[name]
    echo typeof(mm)
    for output in mm.send(src, pulse):
      queue.addLast output

proc part1(input: string): int =
  let text = readFile(input)
  var machine = newMachine(text)

  let pulses = machine.pushButton
  pulses[High] * pulses[Low]

###################

proc timed[T](f: () -> T): T =
  let start = cpuTime()
  result = f()
  echo "Time :", cpuTime() - start

echo timed(proc(): int = part1("input/test.txt"))
# echo timed(proc(): int = part1("input/day20.txt"))
# echo timed(proc(): int = part2("input/test.txt"))
# echo timed(proc(): int = part2("input/day20.txt"))

File: 2023/day20.rb
# frozen_string_literal: true
# typed: true

require 'sorbet-runtime'
require 'sorbet-struct-comparable'

extend T::Sig

sig { params(cond: T::Boolean, msg: String).void }
def assert(cond, msg)
  raise StandardError, msg unless cond
end

sig { params(a: Integer, b: Integer).returns(Integer) }
def gcd(a, b)
  while b != 0
    a, b = b, a % b
  end
  a
end

sig { params(nums: T::Array[Integer]).returns(Integer) }
def gcd_all(nums)
  if nums.length == 0
    return 0
  end

  result = 1
  nums.each do |n|
    result = gcd(result, T.must(n))
  end
  result
end


sig { params(nums: T::Array[Integer]).returns(Integer) }
def lcm(nums)
  gcd = gcd_all(nums)

  result = 1
  nums.each do |n|
    result *= n / gcd
  end
  result
end

module Day20
  class Pulse < T::Enum
    enums do
      High = new
      Low = new
    end
  end

  class Wire < T::Struct
    include T::Struct::ActsAsComparable

    const :src, String
    const :dst, String
    const :pulse, Pulse
  end

  class Module
    extend T::Sig
    extend T::Helpers

    abstract!

    sig { abstract.returns(String) }
    def name; end

    sig { abstract.returns(T::Array[String]) }
    def dests; end

    sig { abstract.params(source: String, pulse: Pulse).returns(T::Array[Wire]) }
    def handle(source, pulse); end
  end

  class Broadcaster < Module
    extend T::Sig

    sig { override.returns(String) }
    attr_reader :name

    sig { override.returns(T::Array[String]) }
    attr_reader :dests

    sig { params(name: String, dests: T::Array[String]).void }
    def initialize(name, dests)
      @name = name
      @dests = dests
    end

    sig { override.params(source: String, pulse: Pulse).returns(T::Array[Wire]) }
    def handle(source, pulse)
      @dests.map do |dst|
        Wire::new(src: @name, dst: dst, pulse: pulse)
      end
    end
  end

  class FlipFlop < Module
    extend T::Sig

    sig { override.returns(String) }
    attr_reader :name

    sig { override.returns(T::Array[String]) }
    attr_reader :dests

    sig { params(name: String, dests: T::Array[String]).void }
    def initialize(name, dests)
      @name = name
      @dests = dests
      @inputs = T.let([], T::Array[String])
      @on = false
    end

    sig { override.params(source: String, pulse: Pulse).returns(T::Array[Wire]) }
    def handle(source, pulse)
      return [] if pulse == Pulse::High

      if @on
        @on = false
        output = Pulse::Low
      else
        @on = true
        output = Pulse::High
      end

      @dests.map do |dst|
        Wire::new(src: @name, dst: dst, pulse: output)
      end
    end
  end

  class Conjunction < Module
    extend T::Sig

    sig { override.returns(String) }
    attr_reader :name

    sig { override.returns(T::Array[String]) }
    attr_reader :dests

    sig { override.params(name: String, dests: T::Array[String]).void }
    def initialize(name, dests)
      @name = name
      @dests = dests
      @inputs = T.let({}, T::Hash[String, Pulse])
    end

    sig { params(name: String).void }
    def register_input(name)
      @inputs[name] = Pulse::Low
    end

    sig { override.params(source: String, pulse: Pulse).returns(T::Array[Wire]) }
    def handle(source, pulse)
      @inputs[source] = pulse

      output = if @inputs.all? { |k, v| v == Pulse::High }
        Pulse::Low
      else
        Pulse::High
      end

      @dests.map do |dst|
        Wire::new(src: @name, dst: dst, pulse: output)
      end
    end
  end

  class Machine
    extend T::Sig

    def initialize(text)
      @modules = T.let({}, T::Hash[String, Module])

      text.strip.split("\n").each do |line|
        name, dests = line.split('->').map(&:strip)
        dests = dests.split(',').map(&:strip)

        mod = if name[0] == '%'
          FlipFlop::new(name[1..name.length], dests)
        elsif name[0] == '&'
          Conjunction::new(name[1..name.length], dests)
        elsif name == 'broadcaster'
          Broadcaster::new(name, dests)
        else
          raise StandardError, name
        end

        @modules[mod.name] = mod
      end

      @modules.each_pair do |src_name, src_mod|
        src_mod.dests.each do |dst_name|
          dst = @modules[dst_name]
          if dst.is_a? Conjunction
            dst.register_input(src_name)
          end
        end
      end
    end

    sig { returns(T::Hash[Pulse, Integer] )}
    def push_button
      counts = {
        Pulse::Low => 0,
        Pulse::High => 0,
      }

      queue = T.let([], T::Array[Wire])
      queue << Wire::new(src: 'button', dst: 'broadcaster', pulse: Pulse::Low)

      while !queue.empty?
        wire = T.must(queue.shift)
        counts[wire.pulse] += 1

        mod = @modules[wire.dst] or next
        mod.handle(wire.src, wire.pulse).each do |output|
          queue << output
        end
      end

      counts
    end

    sig { returns(T::Array[Integer])}
    def run
      gate_firsts = T.let({}, T::Hash[String, Integer])

      count = 0
      while true
        count += 1

        queue = T.let([], T::Array[Wire])
        queue << Wire::new(src: 'button', dst: 'broadcaster', pulse: Pulse::Low)

        while !queue.empty?
          wire = T.must(queue.shift)

          if wire.pulse == Pulse::High and @gates.any?(wire.src)
            gate_firsts[wire.src] = gate_firsts[wire.src] || count
          end

          mod = @modules[wire.dst] or next
          mod.handle(wire.src, wire.pulse).each do |output|
            queue << output
          end
        end

        if @gates.all? { |g| !gate_firsts[g].nil? }
          return gate_firsts.values
        end
      end
    end

    def assert_conditions
      tx = []
      @modules.each_pair do |name, mod|
        tx << name if mod.dests.any?('rx')
      end

      case tx.length
      when 1
        @tx = tx[0]
      when 0
        raise StandardError, 'no transmitter'
      else
        pp tx
        raise StandardError, 'more than one transmitter'
      end

      @gates = []
      @modules.each_pair do |name, mod|
        if mod.dests.any?(@tx)
          @gates << name
          assert mod.dests.length == 1, 'Gate emits to non-tx'
        end
      end

      @exits = []
      @modules.each_pair do |name, mod|
        ix = mod.dests & @gates
        next if ix.empty?

        @exits << name
      end
    end

    sig { params(gv: String, png: String).void }
    def show_graph(gv, png)
      File.open(gv, 'w') do |f|
        f.puts('strict digraph Day20 {')
        @modules.each_pair do |name, mod|
          color = case mod
          when FlipFlop    then '#5FD65D'
          when Conjunction then '#00D7FF'
          else if name == 'rx' then '#FF875F' else 'black' end
          end
          label = T.must(mod.class.name).split('::').last
          exit = if @exits.any?(name) then '*' else '' end

          f.puts("#{name} [fontcolor=\"#{color}\",label=\"\\N\\n#{exit}#{label}\"];")
          mod.dests.each do |dst|
            f.puts("#{name} -> #{dst};")
          end
        end
        f.puts('}')
      end

      system('dot', '-Tpng', gv, out: png)
      # system('imv', png)
    end

    sig { returns(T::Hash[String, T.any(T::Boolean, T::Hash[String, Pulse])])}
    def state
      state = {}
      @modules.each_pair do |name, mod|

      end
      state
    end
  end

  class << self
    extend T::Sig

    sig { params(path: String).returns(Integer) }
    def part1(path)
      text = File.read(path)
      machine = Machine::new(text)
      low = high = 0
      1_000.times do
        pulses = machine.push_button
        high += pulses.fetch(Pulse::High)
        low += pulses.fetch(Pulse::Low)
      end
      high * low
    end

    sig { params(path: String).returns(Integer) }
    def part2(path)
      text = File.read(path)
      machine = Machine::new(text)

      machine.assert_conditions

      gv  = File.join(File.dirname(path), "day20.gv")
      png = File.join(File.dirname(path), "day20.png")
      machine.show_graph(gv, png)
      counts = machine.run
      lcm(counts)
    end
  end
end

# puts Day20::part1 File.join(__dir__, 'input/test.txt')
puts Day20::part1 File.join(__dir__, 'input/day20.txt')
puts Day20::part2 File.join(__dir__, 'input/day20.txt')
